README.md
# ivi-homescreen


IVI Homescreen for Wayland

* Strongly Typed (C++)
* Lightweight
* Source runs on Desktop and Yocto Linux
  * Ubuntu 18+
  * Fedora 33+
  * Yocto Dunfell+
* Platform Channels enabled/disabled via CMake
* OpenGL Texture Framework
* Compositor Sub-surface/Region support
* Vulkan / EGL backend support

# Sanitizer Support

You can enable the sanitizers with SANITIZE_ADDRESS, SANITIZE_MEMORY, SANITIZE_THREAD or SANITIZE_UNDEFINED options in your CMake configuration. You can do this by passing e.g. -DSANITIZE_ADDRESS=On on your command line.

If sanitizers are supported by your compiler, the specified targets will be build with sanitizer support. If your compiler has no sanitizing capabilities you'll get a warning but CMake will continue processing and sanitizing will simply just be ignored.

# Backend Support

### EGL Backend
This is the default build configuration.  To manually build EGL Backend use
```
-DBUILD_BACKEND_WAYLAND_EGL=ON -DBUILD_BACKEND_WAYLAND_VULKAN=OFF
```

### Vulkan Backend
To build Vulkan Backend use
```
-DBUILD_BACKEND_WAYLAND_EGL=OFF -DBUILD_BACKEND_WAYLAND_VULKAN=ON
```

Running Vulkan requires an engine version that supports Vulkan.  Stable does not yet support Vulkan.

# Bundle File Override Logic

If an override file is not present, it gets loaded from default location.

## Optional override files

### icudtl.dat

Bundle Override

    {bundle path}/data/icudtl.dat

Yocto Default

    /usr/share/flutter/icudtl.dat

Desktop Default

    /usr/local/share/flutter/icudtl.dat

### libflutter_engine.so

Bundle Override

    {bundle path}/lib/libflutter_engine.so

Yocto/Desktop Default - https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html

# Command Line Options

`--a={int value}` - Sets the Engine's initial state of Accessibility Feature support.  Requires an integer value.

`--c` - Disables the cursor.

`--d` - Outputs backend debug information.  If Vulkan and Validation Layer is available, it will be loaded.

`--f` - Sets window to fullscreen.

`--w={int value}` - Sets View width.  Requires an integer value.

`--h={int value}` - Sets View height.  Requires an integer value.

`--i={int value}` - Sets ivi-shell surface ID.  Requires an integer value.

`--p={int value}` - Sets Pixel Ratio.  Requires a double value.

`--t={String}` - Sets cursor theme to load.  e.g. --t=DMZ-White

`--b={path to folder}` - Sets the Bundle Path.  A bundle path expects the following folder structure:
```
  Flutter Application (bundle folder)
    data/flutter_assets
    data/icudtl.dat (optional - overrides system path)
    lib/libapp.so
    lib/libflutter_engine.so (optional - overrides system path)
```
* `--j=` - Sets the JSON configuration file.
* `--wayland-event-mask` - Sets events to ignore. e.g. --wayland-event-mask=pointer-axis, or --wayland-event-mask="pointer-axis, touch"

  * Available parameters are:
      pointer, pointer-axis, pointer-buttons, pointer-motion, keyboard, touch

* Dart VM arguments - any additional command line arguments not handled get directly passed to the Dart VM instance.

### JSON Configuration keys

#### Global

`app_id` - Sets Application ID.  Currently only the primary index app_id value is used.

`cursor_theme` - Sets cursor theme to use.  This only applies to command line, and global parameter options.

`disable_cursor` - Disables the cursor.  This only applies to command line, and global parameter options.

`debug_backend` - Enables Backend Debug logic.

`wayland_event_mask` - See command line option --wayland-event-mask

#### View Specific

`view` - Minimum required.  Can be single object or array.

`bundle_path` - sets the Bundle Path.

`window_type` - Currently used for AGL Compositor Window Types.  If not running on AGL compositor,
it will create borderless windows in no particular position.

`width` - sets View width.  Requires an integer value.

`height` - sets View height.  Requires an integer value.

`accessibility_features` - Bitmask of Engine Accessibility Features.  Requires an integer.  See flutter_embedder.h for valid values.

`vm_args` - Array of strings which get passed to the VM instance as command line arguments.

`fullscreen` - Sets window to fullscreen.

`ivi_surface_id` - Sets ivi-shell surface ID.

`fps_output_console` - Setting to `1` FPS count is output to stdout.

`fps_output_overlay` - If `"fps_output_console"=1` and `"fps_output_overlay"=1` the screen overlay is enabled.

`fps_output_frequency` - Optional for FPS.  Changing value controls the update interval.

Minimum definition when using `--j=`
```
{"view":{}}
```

If you used this minimum definition, invocation would look something like this
```
homescreen --j=/tmp/min_cfg.json --b={bundle path} --h={view height} --w={view width}
``` 

### JSON Configuration Example 1

Loads Two Views
1. Gallery app to a 1920x1280 Background window, passing two arguments to the Dart VM
2. Video player to Left Panel sized 320x240 with all accessibility features enabled.
```
/tmp/bg_left_rel.json

{
   "view":[
      {
         "bundle_path":"/home/joel/development/gallery/.homescreen/x86/release",
         "vm_args":["--enable-asserts", "--pause-isolates-on-start"],
         "window_type":"BG",
         "width":1920,
         "height":1280
      },
      {
         "bundle_path":"/home/joel/development/plugins/packages/video_player/video_player/example/.homescreen/x86/release",
         "window_type":"PANEL_LEFT",
         "width":320,
         "height":240,
         "accessibility_features":31
      }
   ]
}

homescreen --j=/tmp/bg_left_rel.json
```

### JSON Configuration Example 2

Loads Single View
1. Fullscreen Gallery app, cursor disabled, backend debug enabled, passing `vm_args` values to the Dart VM
```
/tmp/bg_dbg.json

{
   "disable_cursor":true,
   "debug_backend":true,
   "accessibility_features":31,
   "view":{
      "bundle_path":"/home/joel/development/gallery/.homescreen/x86/release",
      "vm_args":["--no-serve-devtools"],
      "width":1920,
      "height":1280,
      "fullscreen":true
   }
}

homescreen --j=/tmp/bg_dbg.json
```

### Parameter loading order
Only VM Command Line arguments are additive.  Meaning all instances of VM command line references will get added
together; JSON view + JSON global + CLI args.

All other parameters get assigned using the following ordering:

1. JSON Configuration View object parameters
2. JSON Configuration Global (non-view) parameters
3. Command Line parameters (Overrides View and Global parameters)

# CMake Build flags

`ENABLE_XDG_CLIENT` - Enable XDG Client.  Defaults to ON

`ENABLE_AGL_CLIENT` - Enable AGL Client.  Defaults to OFF

`ENABLE_IVI_SHELL_CLIENT` - Enable ivi-shell Client.  Defaults to OFF

`DISABLE_FLUTTER_LOG_MESSAGES` - Disable Flutter Engine Log output.  Defaults to OFF

`BUILD_BACKEND_WAYLAND_EGL` - Build Backend for EGL.  Defaults to ON

`BUILD_EGL_TRANSPARENCY` - Build with EGL Transparency Enabled.  Defaults to ON

`BUILD_BACKEND_WAYLAND_VULKAN` - Build Backend for Vulkan.  Defaults to ON

`BUILD_BACKEND_WAYLAND_DRM` - Build Backend Wayland DRM.  Defaults to OFF

`BUILD_TEXTURE_EGL` - Include EGL Textures.  Defaults to ON

`BUILD_TEXTURE_TEST_EGL` - Includes Test Texture.  Defaults to OFF

`BUILD_TEXTURE_NAVI_RENDER_EGL` - Includes Navi Texture.  Defaults to ON

`BUILD_PLUGIN_ISOLATE` - Include Isolate Plugin.  Defaults to ON

`BUILD_PLUGIN_RESTORATION` - Include Restoration Plugin.  Defaults to ON

`BUILD_PLUGIN_PLATFORM` - Include Platform Plugin.  Defaults to ON

`BUILD_PLUGIN_MOUSE_CURSOR` - Include Mouse Cursor Plugin.  Defaults to ON

`BUILD_PLUGIN_GSTREAMER_EGL` - Include GStreamer Plugin.  Defaults to OFF

`BUILD_PLUGIN_TEXT_INPUT` - Includes Text Input Plugin.  Defaults to ON

`BUILD_PLUGIN_KEY_EVENT` - Includes Key Event Plugin.  Defaults to ON

`BUILD_PLUGIN_URL_LAUNCHER` - Includes URL Launcher Plugin.  Defaults to ON

`BUILD_PLUGIN_PACKAGE_INFO` - Include PackageInfo Plugin.  Defaults to ON

`BUILD_PLUGIN_COMP_SURF` - Include Compositor Surface Plugin.  Defaults to ON

`BUILD_PLUGIN_COMP_REGION` - Include Compositor Region Plugin.  Defaults to ON

`BUILD_PLUGIN_OPENGL_TEXTURE` - Includes OpenGL Texture Plugin.  Defaults to ON 

`BUILD_PLUGIN_NAVIGATION` - Includes Navigation Plugin.  Defaults to ON

`BUILD_PLUGIN_ACCESSIBILITY` - Includes Accessibility Plugin.  Defaults to ON

`BUILD_PLUGIN_PLATFORM_VIEW` - Includes PlatformView Plugin.  Defaults to OFF

`BUILD_PLUGIN_DESKTOP_WINDOW` - Includes Desktop Window Plugin.  Defaults to ON

`BUILD_PLUGIN_SECURE_STORAGE` - Includes Flutter Secure Storage.  Defaults to OFF


_**Backend selections (Vulkan, EGL/GLESv2) are mutually exclusive by design.**_

# x86_64 Desktop development notes

## NVidia GL errors

Running EGL backend on a Lenovo Thinkpad with NVidia drivers may generate many GL runtime errors.
This should resolve it:
```
export __EGL_VENDOR_LIBRARY_FILENAMES=/usr/share/glvnd/egl_vendor.d/50_mesa.json
```

## Ubuntu 16-18

### Logging in

Log out if logged in Login screen
Click on username field
Right-click on the gear icon below username field, and select "Ubuntu on Wayland"
Enter password and login

## Ubuntu 20+ / Fedora 33+

Defaults to Wayland, no need to do anything special

# Build steps

## Required Packages

    sudo add-apt-repository ppa:kisak/kisak-mesa
    sudo apt-get update -y
    sudo apt-get -y install libwayland-dev wayland-protocols \
    mesa-common-dev libegl1-mesa-dev libgles2-mesa-dev mesa-utils \
    libxkbcommon-dev

## Optional Packages

    # To build doxygen documentation
    sudo apt-get -y install doxygen

## GCC/libstdc++ Build

    git clone https://github.com/toyota-connected-na/ivi-homescreen.git
    mkdir build && cd build
    cmake .. -DCMAKE_STAGING_PREFIX=`pwd`/out/usr/local
    make install -j

## Clang/libc++ Build

    git clone https://github.com/toyota-connected-na/ivi-homescreen.git
    mkdir build && cd build
    CC=/usr/lib/llvm-12/bin/clang CXX=/usr/lib/llvm-12/bin/clang++ cmake .. -DCMAKE_STAGING_PREFIX=`pwd`/out/usr/local
    make install -j

### Clang Toolchain Setup

    wget https://apt.llvm.org/llvm.sh
    chmod +x llvm.sh
    sudo ./llvm.sh 12
    sudo apt-get install -y libc++-12-dev libc++abi-12-dev libunwind-dev

## CI Example

    https://github.com/toyota-connected-na/ivi-homescreen/blob/main/.github/workflows/ivi-homescreen-linux.yml

## Debian Package

    make package -j
    sudo apt install ./ivi-homescreen-1.0.0-Release-beta-Linux-x86_64.deb

# Flutter Application
## Build

Confirm flutter/bin is in the path using: `flutter doctor -v`

    cd ~/development/my_flutter_app
    flutter channel beta
    flutter upgrade
    flutter config --enable-linux-desktop
    flutter create .
    flutter build bundle

## Install

loading path for application is: `/usr/local/share/homescreen/bundle`

This is used to leverage symlinks.  Such as:

    cd /usr/local/share/homescreen
    sudo rm -rf bundle
    sudo ln -sf ~/development/my_flutter_app/build/ bundle

Or

    sudo mkdir -p /usr/local/share/homescreen/my_flutter_app/
    sudo cp -r build/* /usr/local/share/homescreen/my_flutter_app/
    sudo ln -sf /usr/local/share/homescreen/my_flutter_app/ bundle

## Running on desktop

Copy a current icudtl.dat to /usr/local/share/flutter
Copy libflutter_engine.so to `/usr/local/lib` or use LD_LIBRARY_PATH to point downloaded engine for build:

    cd <homescreen build>
    export LD_LIBRARY_PATH=`pwd`:$LD_LIBRARY_PATH
    homescreen

## Debug

Setup custom devices to control ivi-homescreen via debugger. 

# CMAKE dependency paths

Path prefix used to determine required files is determined at build.

For desktop `CMAKE_INSTALL_PREFIX` defaults to `/usr/local`
For target Yocto builds `CMAKE_INSTALL_PREFIX` defaults to `/usr`


# Yocto recipes

## Kirkstone

    https://github.com/meta-flutter/meta-flutter/tree/kirkstone/recipes-graphics/toyota

## Dunfell

    https://github.com/meta-flutter/meta-flutter/tree/dunfell/recipes-graphics/toyota


.woven/release-notes.md

# Bug Fixes

* None

# New Features

* None

# Improvements / Comments

* None

# Policy

**- DO NOT ERASE THIS SECTION, EVER! ALSO, FOLLOW IT PLEASE!! -**

## Syntax

All lines in each section shall be edited according to one of the following format:

* None
* [CAN-XXX](https://jira.tmc-stargate.com/browse/CAN-XXX) <summary of the task/story> [\\<change\\>]
* [Internal] Private Note not to share publicly
* If you want to use swear words, make sure to tag the line with the [Private] tag

## Tagging

### Change Type

Tag your change by substituting **\\<change\\>** for:

| Tag       | Change Type                                      |
|-----------|--------------------------------------------------|
| **major** | Change requires a major version number increment |
| **minor** | Change requires a minor version number increment |
| **fix**   | Change required a fix version number increment   |

:pencil: If no **[\\<change\\>]** tag is present, it is assumed that the change is a **[fix]**

### Ignored Release Notes

Every Line containing one of the following Tag will not be used to produce the Public Release notes upon Promotion to Production:

* **[Automatic Deployment]**
* **[Private]**
* **[Internal]**
* **[None]**

:pencil: The match is not case-sensitive!

    

shell/main.cc
// Copyright 2020 Toyota Connected North America
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <algorithm>
#include <csignal>
#include <cstdlib>
#include <sstream>

#include "app.h"
#include "configuration/configuration.h"
#include "logging/logging.h"

#include <flutter/fml/command_line.h>
#include <filesystem>

volatile bool running = true;

/**
 * @brief Signal handler
 * @param[in] signal No use
 * @return void
 * @relation
 * internal
 */
void SignalHandler(int signal) {
  (void)signal;
  DLOG(INFO) << "Ctl+C";
  running = false;
}

/**
 * @brief Check if input is a number
 * @param[in] s String to check if it is a number
 * @return bool
 * @retval true If s is a number
 * @retval false If s is not a number
 * @relation
 * internal
 */
bool IsNumber(const std::string& s) {
  return std::all_of(s.begin(), s.end(),
                     [](char c) { return isdigit(c) != 0; });
}

/**
 * @brief Remove argument from vector
 * @param[in] args Vector of element that matches the argument to be removed
 * @param[in] arg Arguments to be removed
 * @return void
 * @relation
 * internal
 */
void RemoveArgument(std::vector<std::string>& args, const std::string& arg) {
  auto result = std::find(args.begin(), args.end(), arg);
  if (result != args.end()) {
    args.erase(result);
  }
}

/**
 * @brief Main function
 * @param[in] argc Number of arguments
 * @param[in] argv Arguments passed to the program
 * @return int
 * @retval 0 Normal end
 * @retval Non-zero Abnormal end
 * @relation
 * wayland, flutter
 */
int main(int argc, char** argv) {
  struct Configuration::Config config {
    .app_id{}, .json_configuration_path{}, .cursor_theme{}, .disable_cursor{},
        .disable_cursor_set{}, .wayland_event_mask{}, .debug_backend{},
        .debug_backend_set{}, .view{},
  };

  if (Dlt::IsSupported()) {
    // Dlt::Register();
  }

  LOG(INFO) << GIT_BRANCH " @ " GIT_HASH;

  config.view.vm_args.reserve(static_cast<unsigned long>(argc - 1));
  for (int i = 1; i < argc; ++i) {
    config.view.vm_args.emplace_back(argv[i]);
  }

  auto cl = fml::CommandLineFromArgcArgv(argc, argv);

  if (!cl.options().empty()) {
    if (cl.HasOption("j")) {
      cl.GetOptionValue("j", &config.json_configuration_path);
      if (config.json_configuration_path.empty()) {
        LOG(ERROR) << "--j option requires an argument (e.g. "
                      "--j=/tmp/cfg-dbg.json)";
        return EXIT_FAILURE;
      }
      DLOG(INFO) << "Json Configuration: " << config.json_configuration_path;
      RemoveArgument(config.view.vm_args,
                     "--j=" + config.json_configuration_path);
    }
    if (cl.HasOption("a")) {
      std::string accessibility_feature_flag_str;
      cl.GetOptionValue("a", &accessibility_feature_flag_str);
      if (accessibility_feature_flag_str.empty()) {
        LOG(ERROR) << "--a option (Accessibility Features) requires an "
                      "argument (e.g. --a=31)";
        return EXIT_FAILURE;
      }
      int ret = 0;
      try {
        // The following styles are acceptable:
        // 1. decimal: --a=31
        // 2. hex: --a=0x3
        // 3. octet: --a=03
        config.view.accessibility_features =
          static_cast<int32_t>(std::stol(accessibility_feature_flag_str, nullptr, 0));
      } catch (const std::invalid_argument& e) {
        LOG(ERROR)
            << "--a option (Accessibility Features) requires an integer value";
        ret = EXIT_FAILURE;
      } catch (const std::out_of_range& e) {
        LOG(ERROR)
            << "The specified value to --a option, "
            << accessibility_feature_flag_str
            << " is out of range.";
        ret = EXIT_FAILURE;
      }
      if (ret) {
        return ret;
      }
      config.view.accessibility_features =
        Configuration::MaskAccessibilityFeatures(config.view.accessibility_features);
      RemoveArgument(config.view.vm_args,
                     "--a=" + accessibility_feature_flag_str);
    }
    if (cl.HasOption("b")) {
      cl.GetOptionValue("b", &config.view.bundle_path);
      if (config.view.bundle_path.empty() ||
          !std::filesystem::is_directory(config.view.bundle_path)) {
        LOG(ERROR) << "--b (Bundle Path) option requires a directory path "
                      "argument (e.g. "
                      "--b=/usr/share/gallery)";
        return EXIT_FAILURE;
      }
      DLOG(INFO) << "Bundle Path: " << config.view.bundle_path;
      RemoveArgument(config.view.vm_args, "--b=" + config.view.bundle_path);
    }
    if (cl.HasOption("c")) {
      DLOG(INFO) << "Disable Cursor";
      config.disable_cursor = true;
      config.disable_cursor_set = true;
      RemoveArgument(config.view.vm_args, "--c");
    }
    if (cl.HasOption("d")) {
      DLOG(INFO) << "Backend Debug";
      config.debug_backend = true;
      config.debug_backend_set = true;
      RemoveArgument(config.view.vm_args, "--d");
    }
    if (cl.HasOption("f")) {
      DLOG(INFO) << "Fullscreen";
      config.view.fullscreen = true;
      config.view.fullscreen_set = true;
      RemoveArgument(config.view.vm_args, "--f");
    }
    if (cl.HasOption("w")) {
      std::string width_str;
      cl.GetOptionValue("w", &width_str);
      if (!IsNumber(width_str)) {
        LOG(ERROR) << "--w option (Width) requires an integer value";
        return EXIT_FAILURE;
      }
      if (width_str.empty()) {
        LOG(ERROR) << "--w option (Width) requires an argument (e.g. --w=720)";
        return EXIT_FAILURE;
      }
      config.view.width = static_cast<uint32_t>(std::stoul(width_str));
      RemoveArgument(config.view.vm_args, "--w=" + width_str);
    }
    if (cl.HasOption("h")) {
      std::string height_str;
      cl.GetOptionValue("h", &height_str);
      if (!IsNumber(height_str)) {
        LOG(ERROR) << "--h option (Height) requires an integer value";
        return EXIT_FAILURE;
      }
      if (height_str.empty()) {
        LOG(ERROR)
            << "--h option (Height) requires an argument (e.g. --w=1280)";
        return EXIT_FAILURE;
      }
      config.view.height = static_cast<uint32_t>(std::stoul(height_str));
      RemoveArgument(config.view.vm_args, "--h=" + height_str);
    }
    if (cl.HasOption("t")) {
      cl.GetOptionValue("t", &config.cursor_theme);
      if (config.cursor_theme.empty()) {
        LOG(ERROR) << "--t option requires an argument (e.g. --t=DMZ-White)";
        return EXIT_FAILURE;
      }
      DLOG(INFO) << "Cursor Theme: " << config.cursor_theme;
      RemoveArgument(config.view.vm_args, "--t=" + config.cursor_theme);
    }
    if (cl.HasOption("window-type")) {
      cl.GetOptionValue("window-type", &config.view.window_type);
      if (config.view.window_type.empty()) {
        LOG(ERROR) << "--window-type option requires an argument (e.g. "
                      "--window-type=BG)";
        return EXIT_FAILURE;
      }
      DLOG(INFO) << "Window Type: " << config.view.window_type;
      RemoveArgument(config.view.vm_args,
                     "--window-type=" + config.view.window_type);
    }
    if (cl.HasOption("output-index")) {
      std::string output_index_str;
      cl.GetOptionValue("output-index", &output_index_str);
      if (!IsNumber(output_index_str)) {
        LOG(ERROR) << "--output-index option (Wayland Output Index) "
                      "requires an integer value";
        return EXIT_FAILURE;
      }
      if (output_index_str.empty()) {
        LOG(ERROR) << "--output-index option (Wayland Output Index) "
                      "requires an argument (e.g. --output-index=1)";
        return EXIT_FAILURE;
      }
      config.view.wl_output_index =
          static_cast<uint32_t>(std::stoul(output_index_str));
      RemoveArgument(config.view.vm_args, "--output-index=" + output_index_str);
    }
    if (cl.HasOption("xdg-shell-app-id")) {
      cl.GetOptionValue("xdg-shell-app-id", &config.app_id);
      if (config.app_id.empty()) {
        LOG(ERROR) << "--xdg-shell-app-id option requires an argument "
                      "(e.g. --xdg-shell-app-id=gallery)";
        return EXIT_FAILURE;
      }
      DLOG(INFO) << "Application ID: " << config.app_id;
      RemoveArgument(config.view.vm_args,
                     "--xdg-shell-app-id=" + config.app_id);
    }
    if (cl.HasOption("wayland-event-mask")) {
      cl.GetOptionValue("wayland-event-mask", &config.wayland_event_mask);
      if (config.wayland_event_mask.empty()) {
        LOG(ERROR) << "--wayland-event-mask option requires an argument "
                      "(e.g. --wayland-event-mask=pointer-axis,keyboard)";
        return EXIT_FAILURE;
      }
      DLOG(INFO) << "Wayland Event Mask: " << config.wayland_event_mask;
      RemoveArgument(config.view.vm_args,
                     "--wayland-event-mask=" + config.wayland_event_mask);
    }
    if (cl.HasOption("p")) {
      std::string pixel_ratio_str;
      cl.GetOptionValue("p", &pixel_ratio_str);
      if (pixel_ratio_str.empty()) {
        LOG(ERROR) << "--p option (Pixel Ratio) requires an argument "
                      "(e.g. --p=1.1234)";
        return EXIT_FAILURE;
      }

      config.view.pixel_ratio = strtod(pixel_ratio_str.c_str(), nullptr);
      RemoveArgument(config.view.vm_args, "--p=" + pixel_ratio_str);
    }
    if (cl.HasOption("i")) {
      std::string ivi_surface_id_str;
      cl.GetOptionValue("i", &ivi_surface_id_str);
      if (ivi_surface_id_str.empty()) {
        LOG(ERROR) << "--i option (IVI Surface ID) requires an argument "
                      "(e.g. --i=2)";
        return EXIT_FAILURE;
      }

      config.view.ivi_surface_id = static_cast<uint32_t>(
          strtoul(ivi_surface_id_str.c_str(), nullptr, 10));
      RemoveArgument(config.view.vm_args, "--i=" + ivi_surface_id_str);
    }
  }

  auto vm_arg_count = config.view.vm_args.size();
  if (vm_arg_count) {
    DLOG(INFO) << "VM Arg Count: " << vm_arg_count;
    for (auto const& arg : config.view.vm_args) {
      DLOG(INFO) << arg;
    }
  }

  auto configs = Configuration::ParseConfig(config);
  for (auto const& c : configs) {
    Configuration::PrintConfig(c);
  }
  assert(!configs.empty());

  App app(configs);

  std::signal(SIGINT, SignalHandler);

  // run the application
  int ret = 0;
  while (running && ret != -1) {
    ret = app.Loop();
  }

  if (Dlt::IsSupported()) {
    // Dlt::Unregister();
  }

  return EXIT_SUCCESS;
}
